### PR TITLE
fix: silent does not always work (fixes #851)

### LINKED ISSUE #851: silent does not always work
node 8.9.1
shelljs 0.8.1
mac osx

I expect this code to not produce any output but it does. 

```
const shell = require('shelljs');

shell.exec('git rev-parse --abbrev-ref  master2@{upstream}', {silent:true});
```

Run it in a git repository that does not have a branch `master2` and you should see an error printed

```
exec: fatal: no such branch: 'master2'
```

### PR DESCRIPTION
**Pull request by uttpal**

### CODE REVIEW

File: test/exec.js [hunk 1]
```diff
@@ -6,14 +6,20 @@ import test from 'ava';
...
 import shell from '..';
 import utils from './utils/utils';
+import mocks from './utils/mocks';
 
 const CWD = process.cwd();
 const ORIG_EXEC_PATH = shell.config.execPath;
 shell.config.silent = true;
 
+test.beforeEach(() => {
+  mocks.init();
```
Comment (nfischer) [inline test/exec.js]: This line is OK as written. This actually exposes a bug (#862), which I will handle.

Comment (nfischer) [reply test/exec.js]: FYI fix is uploaded (#863). It's OK if this PR lands first.

File: test/exec.js [hunk 2]
```diff
@@ -85,6 +91,14 @@ test('check if stdout + stderr go to output', t => {
   t.is(result.stderr, '1234\n');
 });
 
+test('check if stdout + stderr should not be printed to console if silent', t => {
```
[the remark this answers is not in the document]
Comment (uttpal) [reply test/exec.js]: Above you suggested that `silent` will always be `true`, then there will be no `silent: false`case

[the remark this answers is not in the document]
Comment (nfischer) [reply test/exec.js]: > Above you suggested that silent will always be true

For this PR, correct behavior is to always pass `{ silent: true }` to `common.error()`, regardless of `config.silent` or exec's `{ silent: true }` option.

> then there will be no `silent: false` case?

I think two cases are interesting:

 1. when the user passes `silent: true`, there should be no stdio
 2. when the user passes `silent: false` (or, equivalently, the default), the command's stdio should go to stdio, but exec should not also print `exec: <something here>` to stderr

Does this make sense?

[the remark this answers is not in the document]
Comment (uttpal) [reply test/exec.js]: got it

[the remark this answers is not in the document]
Comment (nfischer) [reply test/exec.js]: @uttpal any update?

**Comment by 3rd-Eden** [comment]
Any update on this? The current work around is to globally silent the module using `sh.config.silent = true` so the build-in `log` function doesn't output anything. It would be nice if this PR could be merged and released.

**Comment by nfischer** [comment]
I believe I still need a change from @uttpal (if that's not the case, please let me know I should re-review this).

File: src/exec.js [hunk 3]
```diff
@@ -96,7 +96,7 @@ function execSync(cmd, opts, pipe) {
   try { common.unlinkSync(stdoutFile); } catch (e) {}
 
   if (code !== 0) {
-    common.error(stderr, code, { continue: true });
+    common.error(stderr, code, { continue: true, silent: opts.silent });
```
[the remark this answers is not in the document]
Comment (uttpal) [reply src/exec.js]: just to clarify, here `silent: true` will always be there?

File: test/exec.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (uttpal) [reply test/exec.js]: I will fix and add test asap

File: src/exec.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (nfischer) [reply src/exec.js]: In my previous comment, I intended for you to change this line to:

```js
common.error(stderr, code, { continue: true, silent: true });
```

**Comment by nfischer** [comment]
FYI I cherry-picked this to #892

**Comment by nfischer** [comment]
Close in favor of #892

**Pull request closed without merging**